src/core/handshaker/security/pipelined_secure_endpoint.cc

```diff
@@ -0,0 +1,938 @@
+            return;
+          }
+
+          // There's more data to unprotect - grab it under the queue lock.
+          source_buffer = std::move(impl->protected_data_buffer_);
+          impl->frame_protector_.TraceOp("data",
+                                         source_buffer->c_slice_buffer());
+          args = std::move(impl->last_read_args_);
+          args.set_read_hint_bytes(impl->frame_protector_.min_progress_size());
+        }
+
+        // Kick off the next read in another thread while we unprotect in this
+        // thread.
+        impl->event_engine_->Run(
+            [impl = impl->Ref(), args = std::move(args)]() mutable {
+              StartAsyncRead(std::move(impl), std::move(args));
+            });
+
+        {
+          grpc_core::ReleasableMutexLock lock(impl->frame_protector_.read_mu());
+          // Unprotect the bytes.
+          CHECK(read_buffer == nullptr);
+          impl->frame_protector_.SetSourceBuffer(std::move(source_buffer));
+          read_buffer = std::make_unique<SliceBuffer>();
+          impl->frame_protector_.BeginRead(read_buffer->c_slice_buffer());
+          unprotect_status = impl->frame_protector_.Unprotect(absl::OkStatus());
+          impl->frame_protector_.FinishRead(unprotect_status.ok());
+          if (!unprotect_status.ok()) {
+            lock.Release();
+            FailReads(std::move(impl), unprotect_status);
+            return;
+          }
+        }
+
+        {
+          grpc_core::ReleasableMutexLock lock(&impl->read_queue_mu_);
+          impl->unprotected_data_buffer_ = std::move(read_buffer);
+          // We have a read waiting on this unprotected data - call the
+          // callback and continue in loop.
+          if (impl->on_read_ != nullptr) {
+            *impl->pending_output_buffer_ =
+                std::move(*impl->unprotected_data_buffer_);
+            impl->unprotected_data_buffer_.reset();
+            on_read = std::move(impl->on_read_);
+            lock.Release();
+            impl->event_engine_->Run([on_read = std::move(on_read)]() mutable {
+              on_read(absl::OkStatus());
+            });
+          } else {
+            // We're waiting on the next read to read this data. We're done
+            // for now.
+            impl->waiting_for_transport_read_ = true;
+            lock.Release();
+            break;
+          }
+        }
+      }
+    }
+
+    static void StartAsyncRead(grpc_core::RefCountedPtr<Impl> impl,
+                               ReadArgs args) {
+      grpc_core::ReleasableMutexLock lock(&impl->shutdown_read_mu_);
+      if (impl->wrapped_ep_ == nullptr) {
+        lock.Release();
+        FailReads(std::move(impl),
+                  absl::CancelledError("secure endpoint shutdown"));
+        return;
+      }
+      const bool read_finished_immediately = impl->wrapped_ep_->Read(
+          [impl = impl->Ref()](absl::Status status) mutable {
+            bool should_unprotect = false;
```

Comment:
I'm wondering if this lambda might be better off as a method - since it's long and interrupts the flow of the enclosing code